tools/hidl/hidlcomp.cpp

```diff
@@ -6608,28 +6602,20 @@ void EspServInfo::write_esp_client()
         outs("\tif(state!=NULL)\n");
         outs("\t\tstate->Link();\n\n");
 
-        
-        outs("#ifdef _WIN32\n");
-        outs("\t_beginthread(espWorkerThread, 0, (void *)(IRpcRequestBinding *)(esprequest));\n");
-        outs("#else\n");
-
-        outs("\tpthread_attr_t attr;\n");
-        outs("\tpthread_attr_init(&attr);\n");
-        outs("\tpthread_attr_setscope(&attr, PTHREAD_SCOPE_SYSTEM);\n");
-        outs("\tpthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);\n");
-        outs("\tpthread_attr_setstacksize(&attr, 0x10000);\n");
-        outs("\tThreadId threadid;\n");
-        outs("\tint status;\n");
-        outs("\tdo\n");
+        outs("\n");
+        outf("\tclass CClient%sThread: public Thread\n", name_);
         outs("\t{\n");
-        outf("\t\tstatus = pthread_create(&threadid, &attr, CClient%s::espWorkerThread, (void *)(IRpcRequestBinding *)(esprequest));\n", name_);
-        outs("\t} while (0 != status && (errno == EINTR));\n");
-        outs("\tif (status) {\n");
-        outs("\t\tRelease();\n");
-        outs("\t\tthrow makeOsException(errno);\n");
-        outs("\t}\n");
-
-        outs("#endif\n");
+        outs("\tpublic:\n");
+        outs("\t\tOwned<IRpcRequestBinding> m_esprequest;\n");
+        outf("\t\tCClient%sThread(IRpcRequestBinding *_esprequest) { m_esprequest.set(_esprequest); }\n", name_);
+        outs("\t\tint run()\n");
+        outs("\t\t{\n");
+        outf("\t\t\tCClient%s::espWorkerThread((void *)m_esprequest);\n", name_);
+        outs("\t\t\treturn 0;\n");
+        outs("\t\t}\n");
+        outs("\t} clientthread(esprequest);\n\n");
```

Comment:
I am surprised this doesn't cause problems.
Thread objects need to be allocated on the heap - if it is on the stack the object will be destroyed as the thread is running, causing it to crash or corrupt memory.

```
Thread * child = new MyThread(data);
child->startRelease();

```

A separate issue, but not for this PR, is that startRelease() is not recommended.  You should always be prepared to join a thread later on - otherwise if you terminate the program before the thread has completed it will crash.  Since esp doesn't tend to terminate that isn't a very significant issue.